UIE-4 Curried Functions - Typescript type and utility helpers

msilva-broad opened the pull request:
Curried Functions - Typescript type and utility helpers

- created SafeCurry2, SafeCurry3 etc types to offer a pattern to type curried functions with stronger type flow then the lodash-provided types.  See lodash-fp-utils.types.ts for usage examples.  The pattern requires functions that will be curried to be defined before being used in lodash .curry() call, but provides better type flow and safety.
- lodash-fp-helpers implements a set of helpers to provide a stronger typed alternative to _.flow usage when adding 'handlers' like withErrorHandling to things like async data calls.  See unit tests for examples.

https://broadworkbench.atlassian.net/jira/software/c/projects/UIE/boards/162?modal=detail&selectedIssue=UIE-4

msilva-broad committed:
UIE-4 Curried Fn type helpers
- initial implementation of type and utility helpers
- SafeCurry2, SafeCurry3 etc offer a pattern to type curried functions with stronger type flow then the lodash-provided types.  See lodash-fp-utils.types.ts for usage examples
- lodash-fp-helpers implements a set of helpers to provide a stronger typed alternative to _.flow usage when adding 'handlers' like withErrorHandling to things like async data calls.

msilva-broad committed:
UIE-4 Curried Fn type helpers
- cleanup

msilva-broad committed:
UIE-4 Curried Fn type helpers
- lint cleanup

nawatts commented on types/lodash-fp/lodash-fp.utils.ts:
```diff
@@ -0,0 +1,22 @@
+// eslint-disable-next-line lodash-fp/use-fp -- we need these types from base lodash lib
```
I think it's unclear what the distinction is between this file and lodash-fp-helpers.ts. And more generally, the intended distinction between types and src/libs/type-utils. Currently, types contains only type declarations for 3rd party code, but this change makes that definition fuzzier. I think I'd suggest merging this into src/libs/lodash-fp-helpers.ts. That would avoid adding any imports from types.

nawatts replied:
Also, it would be nice to have a comment or example explaining the benefit of these over the `@types/lodash` types. Maybe something in the `.types.ts` file showing how `_.curry` allows things it shouldn't?

msilva-broad replied:
both great points. 

Question on the directory question... I would like to keep the check-types validation for the lodash/fp utils.... I assume we are ok with those staying in [root]/tests and just the actual types and utils all living in src/libs/type-utils ?

nawatts replied:
I'm fine with that. In the future, we could possibly revisit how check-dts is run if we want to move the type validation side by side with the modules in src/libs/type-utils.

msilva-broad replied:
I think a big consideration is that .ts files inside the src folder that have errors (expected for check-dts) are a big annoyance in most IDE's that will highlight the directory tree on up with red squigglies to say HEY THERE ARE ERRORS in this folder somewhere...  That might keep us with the very useful check-dts stuff outside of the main src folder.

nawatts commented on src/libs/type-utils/lodash-fp-helpers.test.ts:
```diff
@@ -0,0 +1,451 @@
+      expect(handlersWatcher).toHaveBeenNthCalledWith(2, 'handler2', 'sup', 7)
+    })
+
+    /* top recommendation */
```
Should these tests be reordered to place this at the top of the file where it's most easily findable for reference?

msilva-broad replied:
right now the ordering is basic, to flexible with safety, to safest (recommended), then SafeCurryX for completeness...  I could put SafeCurryX as first example to keep the progression.... the intent was to have the example show the progression and that they build on each other as options.

nawatts replied:
That makes sense. Making this easier to find to reference could also be accomplished with a link in the wiki or something like that.

msilva-broad replied:
most IDE's also let you easily collapse code blocks... which I often do to collapse a set of tests one by one to see the listing more easily.

nawatts commented on src/libs/type-utils/lodash-fp-helpers.test.ts:
```diff
@@ -0,0 +1,451 @@
+
+      // Act
+      // testing curried functions for completeness, but note that they have
+      // inherently less type safety compared to using curryOnlyLastArg or createHandlerAsync helpers
```
> note that they have inherently less type safety compared to using curryOnlyLastArg or createHandlerAsync helpers

Why? Isn't SafeCurry supposed to provide that type safety?

msilva-broad replied:
SafeCurryX provides a general improvement to type safety of curried functions.  The challenge with these handler functions is that they are essentially unknown on the function signature they are wrapping... combine that with the allowed permutations of a curried function (none, some or all args given produces different results), and you get TS able to protect against some, but not all type violations. 

That's better then nothing in this scenario... and SafeCurryX does better in other scenarios.  But this is why the alternatives provided here are recommended over just SafeCurryX when dealing with unknown fn wrappers/handlers.

I can articulate that in a code comment :)

msilva-broad replied:
when implementing your other suggestion to move away from exported function overloads to declared overload interfaces for a couple of the key exported functions in lodash-fp-helpers.ts the type safety on SafeCurryX + withHandlers improved greatly!  So I was able to replace this comment with a more positive one.

nawatts commented on src/libs/type-utils/lodash-fp-helpers.ts:
```diff
@@ -0,0 +1,138 @@
+ * @param handlers
+ * @param mainFn
+ */
+export function withHandlers<P, A extends any[], F extends (...args: A) => Promise<P>>(
```
Given that functions in Terra UI are generally declared with `export const someFn = (...) => { ... }` instead of `export function someFn (...) { ... }`, would it be preferable to define overloads using an interface instead of multiple `export function` statements?

Something like how the types for `h` are declared:
https://github.com/DataBiosphere/terra-ui/blob/c31123e6328f56789203ba3a8f31e60786e61446/types/react-hyperscript-helpers/index.d.ts#L15-L30

msilva-broad replied:
done, with great results!

nawatts commented on src/libs/type-utils/lodash-fp-helpers.ts:
```diff
@@ -0,0 +1,138 @@
+ * handling logic which still honors the same function signature of the fn
+ * it is given as it's single argument
+ * @param handlers
+ * @param mainFn
```
Without descriptions, are these JSDoc parameters useful? Isn't this information provided by the function signature?

msilva-broad replied:
updated

nawatts commented on src/libs/type-utils/lodash-fp-helpers.test.ts:
```diff
@@ -0,0 +1,451 @@
+import { delay } from 'src/libs/utils'
+import { SafeCurry2, SafeCurry3 } from 'types/lodash-fp/lodash-fp.utils'
+
+import { AnyFn, AnyPromiseFn, GenericFn, GenericPromiseFn } from './general-types'
```
All existing Terra UI code uses absolute imports.

msilva-broad replied:
hmm... I poked around a bit on the contributing wiki but couldn't find this callout.... can you point me to it?

I am curious on the reasoning, because I have found that relative paths for 'in same directory or in a sub-directory of focused module area' is very handy in that it makes the code more portable without refactoring import paths.  For things not nearby, yes absolute paths is great in that it calls that out.

As an aside, when we consider shared components/packages, we will need to do extra work to allow for absolute pathing  in those consumable packages so that the consuming packages don't run into problems with the package exports.

nawatts replied:
To my surprise, I also did not find it documented anywhere. 

> I am curious on the reasoning, because I have found that relative paths for 'in same directory or in a sub-directory of focused module area' is very handy in that it makes the code more portable without refactoring import paths. For things not nearby, yes absolute paths is great in that it calls that out.

Yeah, this makes sense to me. I have no objection to relative imports, I only wanted to flag that this was a break with convention in existing code.

nawatts commented on src/libs/type-utils/lodash-fp-helpers.ts:
```diff
@@ -0,0 +1,138 @@
+ * @param fn function with 5 arguments
+ */
+export function curryLastArg<A, B, C, D, LAST, R>(
+  fn: (a: A, b: B, c: C, d:D, last: LAST) => R
```
Formatting nit: space between `d:` and `D`.

msilva-broad replied:
fixed

msilva-broad committed:
UIE-4 Curried Fn type helpers
- initial PR feedback fixes

nawatts commented on src/libs/type-utils/lodash-fp-helpers.ts:
```diff
@@ -0,0 +1,118 @@
+ * currying is removed.
+ * @param handler
+ */
+export const createHandlerAsync = <P, F extends AnyPromiseFn<P>>(
```
What's the benefit of `createHandlerAsync` vs `createHandler`? This is only using `await` in what are effectively `return await` statements. The `awaits` could be removed and this would be the same code as `createHandler`. If the different type signature is helpful, could this be an overload of `createHandler`?

msilva-broad replied:
I toyed around with the overload route, but as I suspected, async comes with distinct expectations that are different from non-asnyc as far as execution order.

createHandlerAsyc expects everything to be async, including the handler fn being passed in.  So everything is properly awaited to match execution order expectations of async.

nawatts replied:
I still don't see how `createAsyncHandler<P, () => Promise<P>>` differs from `createHandler<Promise<P>>` in anything but the type signature.

msilva-broad replied:
execution order.  Calling a method that returns a promise results in a fire-and-forget flow where the caller immediately resumes execution as though the call was instant, but it is off in the background eventually resolving a dangling promise.

While this article is from a related topic, it demonstrates the same root concern:

https://stackoverflow.com/questions/55262996/does-awaiting-a-non-promise-have-any-detectable-effect

nawatts replied:
I'm still skeptical. I don't see a dangling promise here because `result` and `handlerResult` are returned. If `handler` doesn't await `executor` or whatever uses the created the handler doesn't await it, that's on them. It still seems like the difference is the type signatures. I could be wrong though. And in any case, I'm not opposed to a separate function for async handlers, so I won't argue further.

msilva-broad replied:
I think your right, but I had some difficulties wrangling the types until I separated them out... I would like to give it another go, but as a follow-up PR to give space for surprises.  I think some of the other type flow improvements may make this easier to combine into an overload-typed single function.

nawatts commented on src/libs/type-utils/lodash-fp-types.ts:
```diff
@@ -0,0 +1,22 @@
+// provide the desired type-flow through usages, including curried fns used within _.flow
+
+/** more reliable generic typing for curried functions (with 2 args) */
+export type SafeCurry2<F extends (...args: any[]) => any> =
```
Could these use the `AnyFn` type from general-types.ts?

msilva-broad replied:
yes!  AnyFn was a more recent addition.  I will add it's usage here.

msilva-broad replied:
done

nawatts commented on src/libs/type-utils/lodash-fp-helpers.ts:
```diff
@@ -0,0 +1,118 @@
+/*
+The helpers in this file are to provide even better type safety then the baseline lodash-fp utils
+can give when the scenario is known and specific, or a scenario presents additional type-safety
+challenges that the baseline type utils can't fully give type safety for.
```
Formatting nit: other comments in this file have a leading ` * ` on each line.

msilva-broad replied:
fixed

nawatts commented on types/lodash-fp/lodash-fp.utils.errors.ts:
```diff
@@ -0,0 +1,82 @@
+const fnWith2ArgsFn2 = (_x: number[], _y: string) : string => {
+  return 'something interesting'
+}
+const badFn = (_arg1: number): string => {
```
What makes this function "bad"? Or does it only appear in examples of incorrect usage?

msilva-broad replied:
only used in ...errors.ts to force incorrect usage checks

msilva-broad replied:
added comment

nawatts commented on types/lodash-fp/lodash-fp.utils.errors.ts:
```diff
@@ -0,0 +1,82 @@
+  return 'mischief'
+}
+
+const c2Fn1 = _.curry(fnWith2ArgsFn1) as SafeCurry2<typeof fnWith2ArgsFn1>
```
Would it be easier to follow if names like these were spelled out, like `curriedFnWith2ArgsFn1`?

msilva-broad replied:
fixed up the names

nawatts commented on types/lodash-fp/lodash-fp.utils.types.ts:
```diff
@@ -0,0 +1,46 @@
+  return 'something interesting'
+}
+
+const c2Fn1 = _.curry(fnWith2ArgsFn1) as SafeCurry2<typeof fnWith2ArgsFn1>
```
The usage for `SafeCurryN` is pretty verbose and repetitive. Would it be possible to add a helper to simplify this? Like a `curry` function in lodash-fp-helpers.ts overloaded to return the appropriate `SafeCurryN` type?

msilva-broad replied:
I love it!  I had a similar idea floating around in my head.  I will give it a go.

msilva-broad replied:
Hmmm.... This could be tricky...  I am going to recommend this as a follow-up exercise to iterate on what is here.  I'd like to get this in to assist the TS folks, switch to another task, and loop back around in a follow-up PR

nawatts replied:
Sounds good to me.

nawatts reviewed (approved):
I'd still like lodash-fp-types to use the types from general-types and some clearer names in types/lodash-fp files. Other than that, looks good to me.

msilva-broad committed:
UIE-4 Curried Fn type helpers
-  more PR feedback fixes
- improved variable namings in type check files
- brought in AnyFn type usage to lodash-fp-types
- cleaned up file header comment

nawatts commented:
Was there a rebase gone wrong here? GitHub says this PR is +2,603 −161 lines now and the files changed view includes a lot of files that are already present in the dev branch.

msilva-broad commented:
@nick, yes a bad rebase caused some noise.

kyuksel commented on src/libs/type-utils/lodash-fp-helpers.test.ts:
```diff
@@ -0,0 +1,459 @@
+         options are tested and demonstrated to allow greater flexibility if
+         it's needed.
+       - These tests also show recommended type annotations on the
+         handler/wrapper functions
```
Nit:
```suggestion
         handler/wrapper functions.
```

msilva-broad replied:
fixed

kyuksel commented on src/libs/type-utils/lodash-fp-helpers.ts:
```diff
@@ -0,0 +1,118 @@
+}
+
+/**
+ * produces a curried function that expect the last argument to be given
```
Nit:
```suggestion
 * Produces a curried function that expect the last argument to be given
```

Similarly in a few other places

msilva-broad replied:
fixed

kyuksel commented on src/libs/type-utils/lodash-fp-helpers.ts:
```diff
@@ -0,0 +1,118 @@
+ * @param handler
+ */
+export const createHandler = <F extends AnyFn>(
+  handler: (executor: () => ReturnType<F>) => ReturnType<F>
```
A function named `createHandler` taking a `handler` as input doesn't sound very intuitive to me. Should the function be called something like `wrapHandler` or the input be called `fn` or something?

msilva-broad replied:
hmmm... yeah.... I couldn't think of a better name, and not sure if wrapHandler might be more confusing, since the handler arg here is meant to make the handler function read less like a wrapper function and more like a straightforward function that is calling the inner function and reacting.  The createHandler helper is the one doing the wrapping.

So I just left the naming as is, but added an argument description to the jsdoc block.

kyuksel reviewed (approved):
LGTM overall

msilva-broad committed:
UIE-4 curried function type utils
- minor comments cleaup
